src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1650,40 +1650,44 @@ BOOST_AUTO_TEST_CASE(GradientConcatenateLayerTest)
  */
 BOOST_AUTO_TEST_CASE(SimpleLookupLayerTest)
 {
-  arma::mat output, input, delta, gradient;
-  Lookup<> module(10, 5);
+  const size_t vocabSize = 10;
+  const size_t embeddingSize = 2;
+  const size_t seqLength = 3;
+  const size_t batchSize = 4;
+
+  arma::mat output, input, gy, g, gradient;
+
+  Lookup<> module(vocabSize, embeddingSize);
   module.Parameters().randu();
 
   // Test the Forward function.
-  input = arma::zeros(2, 1);
-  input(0) = 1;
-  input(1) = 3;
+  input = arma::zeros(seqLength, batchSize);
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    int token = math::RandInt(1, vocabSize);
+    input(i) = token;
+  }
 
   module.Forward(input, output);
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    // The Lookup module uses index - 1 for the cols.
+    const double outputSum = arma::accu(module.Parameters().cols(
+        arma::conv_to<arma::uvec>::from(input.col(i)) - 1));
 
-  // The Lookup module uses index - 1 for the cols.
-  const double outputSum = arma::accu(module.Parameters().col(0)) +
-      arma::accu(module.Parameters().col(2));
-
-  BOOST_REQUIRE_CLOSE(outputSum, arma::accu(output), 1e-3);
+    BOOST_REQUIRE_CLOSE(outputSum, arma::accu(output.col(i)), 1e-3);
+  }
 
   // Test the Backward function.
-  module.Backward(input, input, delta);
-  BOOST_REQUIRE_EQUAL(arma::accu(input), arma::accu(input));
+  gy = 0.3 * arma::randu(embeddingSize * seqLength, batchSize);
+  module.Backward(input, gy, g);
+  BOOST_REQUIRE_EQUAL(arma::accu(gy), arma::accu(g));
 
   // Test the Gradient function.
-  arma::mat error = arma::ones(2, 5);
-  error = error.t();
-  error.col(1) *= 0.5;
-
+  arma::mat error = 0.01 * arma::randu(embeddingSize * seqLength, batchSize);
   module.Gradient(input, error, gradient);
 
-  // The Lookup module uses index - 1 for the cols.
-  const double gradientSum = arma::accu(gradient.col(0)) +
-      arma::accu(gradient.col(2));
-
-  BOOST_REQUIRE_CLOSE(gradientSum, arma::accu(error), 1e-3);
-  BOOST_REQUIRE_CLOSE(arma::accu(gradient), arma::accu(error), 1e-3);
+  BOOST_CHECK_CLOSE_FRACTION(arma::accu(error), arma::accu(gradient), 1e-05);
```

Comment:
I didn't get this part. Could you explain the purpose of this check?

Reply:
The older test and this test are the same. Actually, I'm not getting your question. Do you suggest another test method I should use? Or your question is about the validity of the `Gradient` function?

Reply:
No, I was speaking about the test. I don't understand the point. What exactly does it check? Does it relevant?

Reply:
Haha... In a sense, you are right because the check is like "what I've done is correct!". But I'm not sure whether we should remove the check or not.
I have just added a gradient test and fortunately, it passed locally.

Reply:
But I'm still wondering what is a better way to check it. I think it's not wrong if you see the `Gradient` function, it can make sense.

Suppose I have a sequence -- "A word embedding is a class of approaches for representing words and documents using a dense vector representation."

Here the word "a" comes three times and so there will be three errors associated with the word "a", we sum up all those errors. And then the gradient here is same as the error, such is this "mathematical function". The gradients are same as errors if there is no repetition of words but if there is repetition, the error for the same words is added and stored as the gradient of that word.

```
grad wrt "a" = sum of errors associated with "a"
grad wrt "word" = sum of errors associated with "word"
grad wrt "embedding" = sum of errors associated with "embedding"
...
```
...and that is why `sum of gradients = sum of errors`. Correct me if I misunderstood something or if I need to clarify anything further.

Reply:
I think the main problem is that `input` is random. So, you pick a random slices from random `error`. The sums are not necessarily equal. Anyway, I think the new test is fine.